### ISSUE TITLE
Support of 'IAsyncComparer' for Linq operations.

### ISSUE THREAD

**Issue by Equijano24** [dotnet/reactive#1676]
Some Linq operations such as OrderBy or ThenBy might require asynchronous comparers for custom objects. Consider the following case:
```
public class ExoticObject {...}

public class ExoticComparer : IAsyncComparer<ExoticObject>
{
    public async ValueTask<int> Compare(ExoticObject e1, ExoticObject e2)
    {
        return await ..... // Some asynchronous code
    }
}
```
The following expression would make use of an asynchronous comparer:
```
IOrderedAsyncEnumerable<ExoticObject> orderedOnes =
    listOfExotics().OrderByAwait(SomeAsyncKeySelector, new ExoticComparer());
```
Please provide an IAsyncComparer interface and support for Linq operations.

**Comment by theodorzoulias** [comment]
Invoking the `Compare` method is expected to be fast, because it is invoked many times during a sorting operation. Your suggestion implies that you are OK with sorting things very slowly. In which case you might as well accept the inefficiency of invoking the asynchronous comparisons synchronously, in other words using the existing sorting infrastructure. You'll block a thread during the sorting, which could be avoided, but how frequently do you have to sort exotic objects anyway?

**Comment by Equijano24** [comment]
You are right about the inefficiency problem. There are two problems that might force having an asynchronous comparer:

1- Having to use third-party packages that have 'async' methods only. Even if you don't really need asynchronous operations you are forced to make your code asynchronous.

2- Working on 'async-only' environments. For example, running code on Blazor with Webassembly (client side) forces you to use async methods without using Task.Result or Task.Wait. You cannot block a thread since Blazor runs on the UI thread only. They have no support for multithreading yet.

**Comment by svick** [comment]
Can you asynchronously get a sort key for each object and then synchronously compare those? If so, would the following code work for you?

```c#
IAsyncEnumerable<ExoticObject> orderedOnes =
    listOfExotics
        .ToAsyncEnumerable()
        .SelectAwait(async o => (o, sortKey: await GetSortKeyAsync(o)))
        .OrderBy(x => x.sortKey)
        .Select(x => x.o);
```

**Comment by Equijano24** [comment]
Using a sort key looks like an awesome solution. It might even be faster than using comparers.
